### PR TITLE
statestore: add initial implementation

### PR DESCRIPTION
**Pull request by acud**
uses leveldb for now. will be converted later to badger

dependency of https://github.com/ethersphere/bee/issues/59

### CODE REVIEW

File: pkg/statestore/store.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	var bytes []byte
+	if marshaler, ok := i.(encoding.BinaryMarshaler); ok {
+		if bytes, err = marshaler.MarshalBinary(); err != nil {
+			return err
+		}
+	} else {
+		if bytes, err = json.Marshal(i); err != nil {
+			return err
+		}
+	}
```
Comment (pradovic) [inline pkg/statestore/store.go:63]:
```suggestion
	} else if bytes, err = json.Marshal(i); err != nil {
		return err
	}
```

**Comment by acud** [comment]
thanks @pradovic.

> I have a feeling that named returned params are a bit overused. I don't think there is a reason to use them when the return params are just error, or error and an expected value like in constructor.

I'm not sure I understand what you mean. Can you please elaborate?

File: pkg/statestore/store.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+// Store uses LevelDB to store values.
+type Store struct {
+	db *leveldb.DB
+}
+
+// New creates a new persistent state storage.
+func New(path string) (s storage.StateStorer, err error) {
```
Comment (pradovic) [inline pkg/statestore/store.go:22]:
```suggestion
func New(path string) (storage.StateStorer, error) {
```

File: pkg/statestore/store.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+		return nil, err
+	}
+	return &Store{
+		db: db,
+	}, nil
+}
+
+// Get retrieves a value of the requested key. If not results are found,
+// storage.ErrNotFound will be returned.
+func (s *Store) Get(key string, i interface{}) (err error) {
```
Comment (pradovic) [inline pkg/statestore/store.go:34]:
```suggestion
func (s *Store) Get(key string, i interface{}) error {
```

**Comment by pradovic** [comment]
@acud I have added a few code suggestions to illustrate the point.

**Comment by acud** [comment]
Thanks @pradovic !

**Pull request merged**